### PR TITLE
use universal Gitea OAuth configuration

### LINKED ISSUE #1426: Automatic Gitea OAuth support
https://github.com/git-ecosystem/git-credential-manager/issues/145#issuecomment-1428774678 explains you can manually configure GCM to authenticate to a Gitea instance by OAuth. Better would be to automatically recognise Gitea instances by WWW-Authenticate header (populated by Gitea 1.21 and later):

> Www-Authenticate: Basic realm="Gitea"

The OAuth configuration is universal across instances: client id [e90ee53c-94e2-48ac-9358-a874fb9e0662 ](https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#pre-configured-applications) , and endpoints `/login/oauth/authorize` and `/login/oauth/access_token`.

Gitea v1.21 instances to test against: https://try.gitea.io/ and https://next.forgejo.org/

### PR DESCRIPTION
**Pull request by hickford**
https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#pre-configured-applications

Tested with https://try.gitea.io/

Fixes #1426

### CODE REVIEW

File: src/shared/Core/GenericOAuthConfig.cs [hunk 1]
```diff
@@ -1,32 +1,56 @@
...
-            if (!settings.TryGetSetting(
+            if (input.WwwAuth.Any(x => x.Contains("Basic realm=\"Gitea\"")))
+            {
+                trace.WriteLine($"Using universal Gitea OAuth configuration");
+                // https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#pre-configured-applications
+                config.ClientId = "e90ee53c-94e2-48ac-9358-a874fb9e0662";
+                // https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#endpoints
+                authzEndpointUri = new Uri(remoteUri, "/login/oauth/authorize");
+                tokenEndpointUri = new Uri(remoteUri, "/login/oauth/access_token");
+                config.RedirectUri = new Uri("http://127.0.0.1");
```
Comment (mjcheetham) [inline src/shared/Core/GenericOAuthConfig.cs:25]: Could we move some of these string literals to constants, perhaps a static inner-class `WellKnownConfig` or something? I can imagine this set of config growing in the future.

```suggestion
                config.ClientId = WellKnown.GiteaClientId;
                authzEndpointUri = new Uri(remoteUri, WellKnown.GiteaAuthzEndpoint);
                tokenEndpointUri = new Uri(remoteUri, WellKnown.GiteaTokenEndpoint);
                config.RedirectUri = WellKnown.LocalIPv4RedirectUri;
```

```csharp
public static class GenericOAuthConfig
{
    ...

    public static class WellKnown
    {
        // https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#pre-configured-applications
        public const string GiteaClientId = "e90ee53c-94e2-48ac-9358-a874fb9e0662";
        // https://docs.gitea.com/next/development/oauth2-provider?_highlight=oauth#endpoints
        public const string GiteaAuthzEndpoint = "/login/oauth/authorize";
        public const string GiteaTokenEndpoint = "/login/oauth/access_token";
        public static Uri LocalIPv4RedirectUri = new Uri("http://127.0.0.1");
    }
}
```

Can also reference these magic strings in the unit tests too.

`GenericOAuthConfig.WellKnown.*`

File: src/shared/Core.Tests/GenericOAuthConfigTests.cs [hunk 2]
```diff
@@ -57,5 +65,43 @@ public void GenericOAuthConfig_TryGet_Valid_ReturnsTrue()
             Assert.Equal(expectedUserName, config.DefaultUserName);
             Assert.True(config.UseAuthHeader);
         }
+
+        [Fact]
+        public void GenericOAuthConfig_TryGet_Gitea()
```
Comment (mjcheetham) [inline src/shared/Core.Tests/GenericOAuthConfigTests.cs:70]: Tests! ❤️

**Review by mjcheetham — CHANGES_REQUESTED** [review]
Great to see Gitea has a universal app registration for GCM (and credential-oauth)!

Just two comments on structure/style of constants and case matching.

**Review by ldennington — APPROVED** [review]

**Review by MissLov3ly — COMMENTED** [review]
.

**Comment by Backuser617** [comment]
I need help

**Pull request merged**